### PR TITLE
Added template + partial render events and a fix for lazyloading images

### PR DESCRIPTION
**Pull request by olivermartinfoster**
#2734

All template and partial `img` tags will now have the `loading="eager"` added by default unless the tag already contains a `loading` attribute. `imageready` will now ignore `loading="lazy"` `img` tags.
 
#### Added
* Introduced `template:preRender`, `template:postRender`, `partial:preRender` and `partial:postRender` events to Adapt
* Sub folder `core/js/fixes` for essential fixes
* `img.lazyload` fix, to inject `loading="eager"` attribute into `img` tags if `loading` attribute is missing
* Added `_fixes` object to config.json
* Added `_fixes._imgLazyLoad: true` to config.json to control image lazy load fix, defaults to true

#### Changed
* `imageready` `background-image` loader to ensure the `loading="eager"` attribute is applied
* `imageready` skips loading checks on explicitly `loading="lazy"` images

### CODE REVIEW

File: src/core/js/fixes/img.lazyload.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      // No img tags found
+      return;
+    }
+    event.value = imgs.reduce((value, img) => {
+      // Check if the img tag already has a loading attribute
+      if (/ loading=/gi.test(img)) {
+        return value;
+      }
+      // Add loading="eager" by default
+      return value.replace(img, img.replace(/<img([^>]*)>/gi, '<img loading="eager"$1>'));
```
Comment (cahirodoherty-learningpool) [inline src/core/js/fixes/img.lazyload.js:26]: Minor: This Regex could be cached and reused

**Comment by olivermartinfoster** [comment]
Sorry Cahir. I thought of another use-case. This should work now if someone adds `loading="lazy"` to their template in addition to adding `loading="eager"` by default.

**Comment by olivermartinfoster** [comment]
Ready for review.

File: src/core/js/fixes/img.lazyload.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+define([
+  'core/js/adapt',
+  '../templates'
+], function(Adapt) {
+
+  /**
+   * 27 April 2020 https://github.com/adaptlearning/adapt_framework/issues/2734
+   * Chrome on Android 10 defers load events on images when lite mode is enabled
```
Comment (olivermartinfoster) [inline src/core/js/fixes/img.lazyload.js:8]: remove reference to android version

File: src/core/js/fixes/img.lazyload.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      // No img tags found
+      return;
+    }
+    event.value = imgs.reduce((value, img) => {
+      // Check if the img tag has a loading attribute
+      if (hasLoadingAttr.test(img)) {
+        return value;
+      }
+      // Add loading="eager" by default
+      return value.replace(img, img.replace(findImgTag, '<img loading="eager"$1>'));
```
Comment (olivermartinfoster) [inline src/core/js/fixes/img.lazyload.js:28]: add default in config.json

**Comment by olivermartinfoster** [comment]
Read to review

File: src/core/schema/config.model.schema [hunk 4]
```diff
@@ -176,6 +176,17 @@
         }
       }
     },
+    "_fixes": {
+      "type": "object",
+      "default": {},
```
Comment (tomgreenfield) [inline src/core/schema/config.model.schema:181]: Should the default be this?
```json
{
  "_imgLazyLoad": true
}
```

Comment (olivermartinfoster) [reply src/core/schema/config.model.schema:181]: Yes

Comment (olivermartinfoster) [reply src/core/schema/config.model.schema:181]: Yes

File: src/core/js/fixes/img.lazyload.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    applyImgLoadingFix();
+  });
+
+  function applyImgLoadingFix() {
+    const findImgTag = /<img([^>]*)>/gi;
+    const hasLoadingAttr = / loading=/gi;
+    Adapt.on('template:postRender partial:postRender', function(event) {
+      const imgs = event.value.match(findImgTag);
+      if (!imgs) {
+        // No img tags found
```
Comment (moloko) [inline src/core/js/fixes/img.lazyload.js:26]: comment seems a bit unnecessary... if you feel it needs clarity maybe rename variable to `imgTagsFound`? but I think it's clear enough already

File: src/core/js/fixes/img.lazyload.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        return;
+      }
+      event.value = imgs.reduce((value, img) => {
+        // Check if the img tag has a loading attribute
```
Comment (moloko) [inline src/core/js/fixes/img.lazyload.js:30]: this comment also seems unnecessary

**Pull request merged**